### PR TITLE
Add py.test

### PR DESCRIPTION
**Pull request by WyseNynja**
http://pytest.org/latest/ is really nice for writing tests

I moved over the tests in enc_wrapper just as a starting point.

What do you think? I'm happy to move the current tests over to pytest if you like it.

```
$ PYTHONPATH=.:$PYTHONPATH py.test
================================================= test session starts ==================================================
platform darwin -- Python 2.7.10, pytest-2.8.5, py-1.4.31, pluggy-0.3.1
rootdir: /Users/bwstitt/code/joinmarket, inifile: setup.cfg
plugins: cov-2.2.0
collected 4 items 

test/test_enc_wrapper.py ....
----------------------------------- coverage: platform darwin, python 2.7.10-final-0 -----------------------------------
Name                                Stmts   Miss  Cover
-------------------------------------------------------
joinmarket/__init__.py                 21      4    81%
joinmarket/blockchaininterface.py     488    422    14%
joinmarket/configure.py               113     71    37%
joinmarket/enc_wrapper.py              26      6    77%
joinmarket/irc.py                     481    427    11%
joinmarket/jsonrpc.py                  48     36    25%
joinmarket/maker.py                   254    214    16%
joinmarket/message_channel.py          68     47    31%
joinmarket/old_mnemonic.py             30     25    17%
joinmarket/slowaes.py                 357    318    11%
joinmarket/socks.py                   194    153    21%
joinmarket/support.py                 214    183    14%
joinmarket/taker.py                   359    308    14%
joinmarket/wallet.py                  238    193    19%
test/commontest.py                     50     50     0%
test/fee-estimation.py                 90     90     0%
test/randomfunc-test.py                59     59     0%
test/regtest.py                        81     81     0%
test/test_enc_wrapper.py               26      0   100%
test/tumbler-test.py                   72     72     0%
test/wallet-test.py                    84     84     0%
-------------------------------------------------------
TOTAL                                3353   2843    15%

=============================================== 4 passed in 3.20 seconds ============================================
```

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+matplotlib
```
Comment (WyseNynja) [inline requirements-dev.txt]: this is optional. should we install it separately?

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[tox]
```
Comment (WyseNynja) [inline tox.ini]: this is pre-emptive as tox depends on a setup.py, but it is really handy if we want to support multiple versions of python

https://testrun.org/tox/latest/example/pytest.html

Comment (raedah) [inline tox.ini]: we do have plans to move to python3, so this could be helpful

**Comment by raedah** [comment]
What are the instructions for running these test?

**Comment by WyseNynja** [comment]
@raedah I added steps to the readme: `PYTHONPATH=.:$PYTHONPATH py.test`

Once joinmarket is moved to a proper package with a setup.py, the PYTHONPATH part won't be needed anymore.

I'll add `pip install -r requirements-dev.txt` to the instructions too

**Comment by WyseNynja** [comment]
I didn't mean to trim the trailing whitespace, my editor just does that. I can leave those there if you care.

**Comment by AdamISZ** [comment]
@WyseNynja thanks for making this contribution, and sorry there has been no response as yet. One element of uncertainty is that @ghtdak has worked on significant refactoring of the code, part of which involved moving and improving the testing setup (which as you can see is primitive at the moment). 

From a quick look, I'd think there'd be no harming in adding this as a first step, whatever direction we go later. But I'd like to hear others opinions, @chris-belcher @adlai

**Comment by chris-belcher** [comment]
I agree this is a good idea in principle, I haven't been able to try it out myself yet.

**Comment by raedah** [comment]
yes, i could definitely use some unit tests to catch bugs. I'll defer to @ghtdak to get this in though. My understanding is we are moving to his branch soonish.

**Comment by AdamISZ** [comment]
I've revisited this, since we are keen to add tests specifically for certain parts of the bitcoin code.

It seems to work well, but of course since we're a long way forward now, I had to do a manual merge. I'm going to build further steps of testing on top of this commit in a separate PR (since I can't add in my own merge commit here), hope that's OK @WyseNynja ? 

The main tests (the ones called regtest and tumbler-test which test end to end functioning) are currently using subprocess, which is no good really, so will need some reworking to call the code directly. The other tests in tests/ currently are somewhat trivial, so if you (or anyone) feels like looking at that, please do. I'll focus on taking the transaction serialization tests that @chris-belcher just wrote, adding them in this pytest structure, and extending them to cover as much of the bitcoin code as possible (the stuff we actually use).

**Comment by WyseNynja** [comment]
Sounds good. I don't have the time to contribute right now, so anything to get tests in will make me happy

**Comment by AdamISZ** [comment]
This commit is now in #477, closing this, discussions can continue there.

**Pull request closed without merging**